config/src/main/java/org/springframework/security/config/annotation/web/configurers/saml2/Saml2LogoutConfigurer.java

```diff
@@ -0,0 +1,621 @@
+	 */
+	public Saml2LogoutConfigurer<H> logoutResponse(
+			Customizer<LogoutResponseConfigurer> logoutResponseConfigurerCustomizer) {
+		logoutResponseConfigurerCustomizer.customize(this.logoutResponseConfigurer);
+		return this;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public void configure(H http) throws Exception {
+		RelyingPartyRegistrationResolver relyingPartyRegistrationResolver = getRelyingPartyRegistrationResolver(http);
+		CsrfTokenRepository csrfTokenRepository = http.getSharedObject(CsrfTokenRepository.class);
+		if (csrfTokenRepository != null) {
+			this.logoutHandlers.add(new CsrfLogoutHandler(csrfTokenRepository));
+		}
+		this.logoutHandlers.add(this.contextLogoutHandler);
+		this.logoutHandlers.add(postProcess(new LogoutSuccessEventPublishingLogoutHandler()));
+		LogoutFilter logoutFilter = createLogoutFilter(http, this.logoutHandlers, relyingPartyRegistrationResolver);
+		http.addFilterBefore(logoutFilter, LogoutFilter.class);
+		Saml2LogoutRequestFilter logoutRequestFilter = createLogoutRequestFilter(this.logoutHandlers,
+				relyingPartyRegistrationResolver);
+		http.addFilterBefore(logoutRequestFilter, CsrfFilter.class);
+		Saml2LogoutResponseFilter logoutResponseFilter = createLogoutResponseFilter(relyingPartyRegistrationResolver);
+		logoutResponseFilter.setLogoutSuccessHandler(getLogoutSuccessHandler());
+		http.addFilterBefore(logoutResponseFilter, CsrfFilter.class);
+	}
+
+	/**
+	 * Returns true if the logout success has been customized via
+	 * {@link #logoutSuccessUrl(String)} or
+	 * {@link #logoutSuccessHandler(LogoutSuccessHandler)}.
+	 * @return true if logout success handling has been customized, else false
+	 */
+	boolean isCustomLogoutSuccess() {
+		return this.logoutSuccessHandler != null;
+	}
+
+	private RelyingPartyRegistrationResolver getRelyingPartyRegistrationResolver(H http) {
+		RelyingPartyRegistrationRepository registrations = getRelyingPartyRegistrationRepository();
+		return new DefaultRelyingPartyRegistrationResolver(registrations);
+	}
+
+	private RelyingPartyRegistrationRepository getRelyingPartyRegistrationRepository() {
+		if (this.relyingPartyRegistrationRepository == null) {
+			this.relyingPartyRegistrationRepository = getBeanOrNull(RelyingPartyRegistrationRepository.class);
+		}
+		return this.relyingPartyRegistrationRepository;
+	}
+
+	private LogoutFilter createLogoutFilter(H http, List<LogoutHandler> logoutHandlers,
+			RelyingPartyRegistrationResolver relyingPartyRegistrationResolver) {
+		LogoutHandler[] handlers = logoutHandlers.toArray(new LogoutHandler[0]);
+		LogoutSuccessHandler logoutRequestSuccessHandler = this.logoutRequestConfigurer
+				.logoutRequestSuccessHandler(relyingPartyRegistrationResolver);
+		LogoutSuccessHandler finalSuccessHandler = getLogoutSuccessHandler();
+		LogoutSuccessHandler logoutSuccessHandler = (request, response, authentication) -> {
+			if (authentication == null) {
+				finalSuccessHandler.onLogoutSuccess(request, response, null);
```

Comment:
Nice catch. The `instanceof` check is no longer necessary and is likely a vestige of an earlier iteration, so I'll remove that.

I believe there are some corner cases where `authentication` is `null`. For example, if CSRF is (unwisely) disabled, then if I do `POST /saml2/logout` after the session terminates, then `authentication` will be `null`. I believe `authentication` can be `null` with standard logout as well in the same circumstance.

With standard logout, `LogoutFilter` can still do all the cleanup it needs to and redirect to the login page in this scenario.

With SAML 2.0 logout, `LogoutFilter` can do all the cleanup it needs, but it cannot redirect to the asserting party since `LogoutRequest#NameID` is required. The most reasonable thing seems to be to fall back to the configured ending `LogoutSuccessHandler` as this matches standard logout the closest.

We could avoid an inner class by introducing something like `AuthenticatedRequestMatcher`:

```java
LinkedHashMap<RequestMatcher, LogoutSuccessHandler> successHandlers = new LinkedHashMap<>();
successHandlers.put(new AuthenticatedRequestMatcher(), logoutRequestSuccessHandler);
successHandlers.put(AnyRequestMatcher.INSTANCE, finalSuccessHandler);
DelegatingLogoutSuccessHandler logoutSuccessHandler = new DelegatingLogoutSuccessHandler(successHandlers);
```

I'm open to suggestions, though.

Reply:
I landed on updating `Saml2LogoutRequestSuccessHandler` to include support for when the user is already logged out.

Note also that because of that, the class was also renamed to `Saml2LogoutSuccessHandler`.